networkx/algorithms/bipartite/edge_coloring.py

```diff
@@ -0,0 +1,279 @@
+        # Find a saturating matching with maximum degree using a custom function
+        matching = _matching_saturating_max_degree(G1, top_nodes)
+
+        # Assign colors to the edges in the matching
+        for u, v in list(matching.items()):
+            edge = (u, v)
+            coloring[edge] = i
+
+        # Remove the edges in the matching from G1
+        G1.remove_edges_from(list(matching.items()))
+
+        # Increment the color counter
+        i += 1
+
+    return coloring
+
+
+def _matching_saturating_max_degree(G, top_nodes=None):
+    """Returns a maximum-degree saturating matching in the bipartite graph `graph`.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        The input bipartite graph.
+    top_nodes : list, optional
+        List of nodes that belong to the top node set.
+
+    Returns
+    -------
+    matching : dict
+        A dictionary representing the maximum-degree saturating matching.
+    """
+    degrees = dict(G.degree())
+    max_degree = max(degrees.values())
+    max_degree_vertices = {node for node, deg in degrees.items() if deg == max_degree}
+
+    # two parts of the bipartite graph
+    part_a, part_b = nx.bipartite.sets(G, top_nodes)
+
+    # make two different graphs A_major and B_major
+    A_major_nodes = (part_a & max_degree_vertices) | part_b
+    B_major_nodes = (part_b & max_degree_vertices) | part_a
+
+    A_major = G.subgraph(A_major_nodes)
+    B_major = G.subgraph(B_major_nodes)
+
+    top_nodes_A = top_nodes & A_major_nodes
+    top_nodes_B = top_nodes & B_major_nodes
+
+    M1 = nx.bipartite.maximum_matching(A_major, top_nodes_A)
+    M2 = nx.bipartite.maximum_matching(B_major, top_nodes_B)
+
+    # combine these two matchings
+    final_matching = {}
+    visited = set()
+
+    for u, v in list(M1.items()) + list(M2.items()):
+        if (u, v) in visited:
+            continue
+        u1 = u
+        edges = [(u, v)]
+        visited.update([(u, v), (v, u)])
+        M = M2
+
+        while v in M and v != u1:
+            u, v = v, M[v]
+            edges.append((u, v))
+            visited.update([(u, v), (v, u)])
+            M = M1 if M is M2 else M2
+
+        if v not in M:
+            u, v = v, u
+            edges = [(u, v)]
+            visited.update([(u, v), (v, u)])
+
+            while v in M:
+                u, v = v, M[v]
+                edges.append((u, v))
+                visited.update([(u, v), (v, u)])
+                M = M1 if M is M2 else M2
+
+            u, v = edges[0]
+            if (len(edges) % 2 == 1) or (
+                (u in M1 and M1[u] == v and u in part_a)
+                or (u in M2 and M2[u] == v and u in part_b)
+            ):
+                k = 0
+            else:
+                k = 1
+
+            for u1, v1 in edges[k::2]:
+                final_matching[u1] = v1
+                final_matching[v1] = u1
+        elif v == u1:
+            for i in range(0, len(edges), 2):
+                u, v = edges[i]
```

Comment:
Similar to 4 lines above -- couldn't you use:
```suggestion
            for u, v in edges[::2]:
```